src/iocost/iocost.c

```diff
@@ -0,0 +1,299 @@
+                                        state = SKIP;
+                                        break;
+                                }
+
+                                asprintf(&model, "%u:%u model=linear ctrl=user %s", major(dev), minor(dev), value);
+
+                                state = QOS;
+                                break;
+                        case QOS:
+                                asprintf(&qos, "%u:%u enable=1 ctrl=user %s", major(dev), minor(dev), value);
+
+                                state = DONE;
+                                break;
+                        case SKIP:
+                                state = MODEL;
+                                break;
+                        case DONE:
+                        default:
+                                assert_not_reached();
+                }
+        }
+
+        /* No iocost qos / model parameters found for this device. */
+        if (!name)
+                return 0;
+
+        if (!model || !qos)
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                       "Could not find iocost parameters with name `%s'", name_to_apply);
+
+        printf("Applying `%s' iocost parameters to %s\n\tio.cost.model: %s\n\tio.cost.qos: %s\n", name, path, model, qos);
+
+        r = cg_set_attribute("io", "/", "io.cost.qos", qos);
+        if (r < 0)
+                return log_error_errno(r, "Failed to set qos: %m");
+
+        r = cg_set_attribute("io", "/", "io.cost.model", model);
+        if (r < 0)
+                return log_error_errno(r, "Failed to set model: %m");
+
+        return 0;
+}
+
+static int show_solutions_for_path(const char *path) {
+        _cleanup_free_ char *modalias = NULL;
+        _cleanup_(sd_hwdb_unrefp) sd_hwdb *hwdb = NULL;
+        enum HwDbParseState state = MODEL;
+        const char *key, *value;
+        int r;
+
+        r = hwdb_query_for_path(path, &hwdb, &modalias);
+        if (r < 0)
+                return r;
+
+        printf("Known iocost solutions for %s\n", path);
+        SD_HWDB_FOREACH_PROPERTY(hwdb, modalias, key, value) {
+                _cleanup_free_ char *name = NULL;
+
+                switch (state) {
+                        case MODEL:
+                                name = name_from_key(key);
+                                printf("\n%s:\n\tio.cost.model: %s\n", name, value);
+
+                                state = QOS;
+                                break;
+                        case QOS:
+                                printf("\tio.cost.qos: %s\n", value);
+
+                                state = MODEL;
+                                break;
+                        case DONE:
+                        default:
+                                assert_not_reached();
+                }
+        }
+
+        return 0;
+}
+
+static int validate_path(const char *path) {
+        if (!path_startswith(path, "/sys/") && !path_startswith(path, "/dev/")) {
```

Comment:
`sd_device_new_from_path()` validates the path. Is this really necessary?

Reply:
I wanted a clearer error message, but I don't think it's needed indeed, it should be obvious enough as is.